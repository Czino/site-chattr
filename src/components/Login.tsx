import { atom, useAtom } from 'jotai'
import { useLogin } from 'nostr-hooks'
import { init as initNostrLogin } from 'nostr-login'
import { useEffect } from 'react'
import { Button } from './Button'

export const isLoggedInAtom = atom(false)

export const Login = () => {
    const [, setIsLoggedIn] = useAtom(isLoggedInAtom)
    const { loginWithExtension, logout } = useLogin()

    useEffect(() => {
        initNostrLogin({
            noBanner: false,
            onAuth: () => loginWithExtension({ onSuccess: () => setIsLoggedIn(true) }),
        })
        document.addEventListener('nlAuth', (e: any) => {
            if (e.detail.type === 'login' || e.detail.type === 'signup') {
                loginWithExtension({ onSuccess: () => setIsLoggedIn(true) })
            } else {
                logout()
                setIsLoggedIn(false)
            }
        })
    })

    const launchLogin = () => {
        document.dispatchEvent(new CustomEvent('nlLaunch', { detail: 'welcome' }))
    }
    return (
        <div className="grid grid-cols-1 gap-2 text-center">
            <Button onClick={launchLogin}>Login</Button>
        </div>
    )
}
