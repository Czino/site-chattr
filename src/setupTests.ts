import '@testing-library/jest-dom'
import { toMatchDiffSnapshot } from 'snapshot-diff'
import { TextDecoder, TextEncoder } from 'util'
expect.extend({ toMatchDiffSnapshot })
Object.assign(global, { TextDecoder, TextEncoder })

export const testPubkey = 'c446c97e0b3a2f5e1d33ba40664fa77b03b0afcba9684daf4f7925ae8f5c6fe0'
export const mockLoginFromLocalStorage = jest.fn().mockImplementation(({ onSuccess }) => {
    if (onSuccess) onSuccess()
})
export const mockLoginWithSecretKey = jest.fn().mockImplementation(({ onSuccess }) => {
    if (onSuccess) onSuccess()
})
export const mockLoginWithExtension = jest.fn().mockImplementation(({ onSuccess }) => {
    if (onSuccess) onSuccess()
})
export const mockLogout = jest.fn()
export const mockUseSubscribe = jest.fn().mockReturnValue({
    App: [],
    eose: false,
})
jest.mock('nostr-hooks', () => ({
    useNostrHooks: jest.fn(),
    useLogin: () => ({
        loginFromLocalStorage: mockLoginFromLocalStorage,
        loginWithExtension: mockLoginWithExtension,
        loginWithSecretKey: mockLoginWithSecretKey,
        logout: mockLogout,
    }),
    useActiveUser: () => ({
        activeUser: { pubkey: testPubkey },
    }),
    useSubscribe: (...args: unknown[]) => mockUseSubscribe(...args),
    useNewEvent: () => ({ createNewEvent: jest.fn() }),
}))
