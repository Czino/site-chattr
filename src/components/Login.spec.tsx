import { act, render, renderHook, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { useAtom } from 'jotai'
import { mockLoginWithExtension, mockLogout } from '../setupTests'
import { isLoggedInAtom, Login } from './Login'

const privateKey = '53fb738ea97a408dcf726b9e497600cf79d3a917589a17694aa4a837ef7bdff7'
const mockInit = jest.fn()
jest.mock('nostr-login', () => ({
    init: (...args: unknown[]) => mockInit(...args),
}))

jest.mock('@nostr-dev-kit/ndk', () => ({
    NDKPrivateKeySigner: {
        generate: () => ({
            privateKey,
        }),
    },
}))
describe('Login', () => {
    it('should render correctly', () => {
        const base = render(<Login />).asFragment()

        expect(base).toMatchSnapshot()
    })
    it('should call launch login on button click', async () => {
        const nlLaunch = jest.fn()
        document.addEventListener('nlLaunch', nlLaunch)
        render(<Login />)
        await userEvent.click(screen.getByText('Login'))
        expect(nlLaunch).toHaveBeenCalled()
    })
    it('should call nostr login', () => {
        render(<Login />)
        expect(mockInit).toHaveBeenCalled()
    })
    it('should login with extension on auth', () => {
        render(<Login />)
        act(mockInit.mock.calls[0][0].onAuth)
        expect(mockLoginWithExtension).toHaveBeenCalled()
        const { result } = renderHook(useAtom, { initialProps: isLoggedInAtom })
        expect(result.current[0]).toBeTruthy()
    })
    it('should login with on lnAuth login event', () => {
        render(<Login />)
        document.dispatchEvent(new CustomEvent('nlAuth', { detail: { type: 'login' } }))
        expect(mockLoginWithExtension).toHaveBeenCalled()
        const { result } = renderHook(useAtom, { initialProps: isLoggedInAtom })
        expect(result.current[0]).toBeTruthy()
    })
    it('should login with on lnAuth signup event', () => {
        render(<Login />)
        document.dispatchEvent(new CustomEvent('nlAuth', { detail: { type: 'signup' } }))
        expect(mockLoginWithExtension).toHaveBeenCalled()
        const { result } = renderHook(useAtom, { initialProps: isLoggedInAtom })
        expect(result.current[0]).toBeTruthy()
    })
    it('should logout with on lnAuth logout event', () => {
        render(<Login />)
        const { result } = renderHook(useAtom, { initialProps: isLoggedInAtom })
        // @ts-ignore
        act(() => result.current[1](true))
        expect(result.current[0]).toBeTruthy()
        act(() => document.dispatchEvent(new CustomEvent('nlAuth', { detail: { type: 'logout' } })))
        expect(mockLogout).toHaveBeenCalled()
        expect(result.current[0]).toBeFalsy()
    })
})
