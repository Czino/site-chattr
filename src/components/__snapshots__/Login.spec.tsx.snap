// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correctly 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-1 gap-2 text-center"
  >
    <button
      class="bg-gradient-purple text-white p-4 text-xl uppercase font-bold rounded hover:bg-gradient-purple-dark "
    >
      Login
    </button>
  </div>
</DocumentFragment>
`;
